segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/filesystem/FSStorageTest.java

```diff
@@ -0,0 +1,314 @@
+
+        val si = storage.getStreamSegmentInfo(handle.getSegmentName(), TIMEOUT).join();
+        Assert.assertFalse("Segment was sealed after rejected call to seal.", si.isSealed());
+
+        AssertExtensions.assertThrows(
+                "Delete was allowed on fenced Storage.",
+                () -> storage.delete(handle, TIMEOUT),
+                ex -> ex instanceof StorageNotPrimaryException);
+        boolean exists = storage.exists(handle.getSegmentName(), TIMEOUT).join();
+        Assert.assertTrue("Segment was deleted after rejected call to delete.", exists);
+    }
+
+    /**
+     * Tests the write() method.
+     *
+     * @throws Exception if an unexpected error occurred.
+     */
+    @Override
+    @Test
+    public void testWrite() throws Exception {
+        String segmentName = "foo_write";
+        int appendCount = 100;
+
+        try (Storage s = createStorage()) {
+            s.initialize(DEFAULT_EPOCH);
+            s.create(segmentName, TIMEOUT).join();
+
+            // Invalid handle.
+            val readOnlyHandle = s.openRead(segmentName).join();
+            assertThrows(
+                    "write() did not throw for read-only handle.",
+                    () -> s.write(readOnlyHandle, 0, new ByteArrayInputStream("h".getBytes()), 1, TIMEOUT),
+                    ex -> ex instanceof IllegalArgumentException);
+
+            assertThrows(
+                    "write() did not throw for handle pointing to inexistent segment.",
+                    () -> s.write(createHandle(segmentName + "_1", false, DEFAULT_EPOCH), 0, new ByteArrayInputStream("h".getBytes()), 1, TIMEOUT),
+                    ex -> ex instanceof StreamSegmentNotExistsException);
+
+            val writeHandle = s.openWrite(segmentName).join();
+            long offset = 0;
+            for (int j = 0; j < appendCount; j++) {
+                byte[] writeData = String.format("Segment_%s_Append_%d", segmentName, j).getBytes();
+                ByteArrayInputStream dataStream = new ByteArrayInputStream(writeData);
+                s.write(writeHandle, offset, dataStream, writeData.length, TIMEOUT).join();
+                offset += writeData.length;
+            }
+
+            // Check bad offset.
+            final long finalOffset = offset;
+            assertThrows("write() did not throw bad offset write (larger).",
+                    () -> s.write(writeHandle, finalOffset + 1, new ByteArrayInputStream("h".getBytes()), 1, TIMEOUT),
+                    ex -> ex instanceof BadOffsetException);
+
+            // Check post-delete write.
+            s.delete(writeHandle, TIMEOUT).join();
+            assertThrows("write() did not throw for a deleted StreamSegment.",
+                    () -> s.write(writeHandle, 0, new ByteArrayInputStream(new byte[1]), 1, TIMEOUT),
+                    ex -> ex instanceof StreamSegmentNotExistsException);
+        }
+    }
+
+    //region synchronization unit tests
+    @Test
```

Comment:
JavaDoc explaining in some level of detail what the test does, and what scenarios it covers.